quill/api/src/query.rs

```diff
@@ -1,19 +1,28 @@
 //! Query for all entities with a certain set of components.
 
-use std::{marker::PhantomData, mem::MaybeUninit};
+use std::{
+    marker::PhantomData,
+    mem::MaybeUninit,
+    ops::{Deref, DerefMut},
+};
 
 use quill_common::{entity::QueryData, Component, HostComponent, PointerMut};
 
+use itertools::Itertools;
+
 use crate::{Entity, EntityId};
 
 /// A type that can be used for a query.
 ///
 /// Implemented for tuples of `Query`s as well.
 pub trait Query {
     type Item;
+    type Target;
 
     fn add_component_types(types: &mut Vec<HostComponent>);
 
+    fn borrowed_mut(ty: HostComponent) -> bool;
```

Comment:
This can be simplified to
```rust
const MUT: bool = true;
const MUT: bool = false;
```

Reply:
I don't think this can be, see the tuple query impl

Reply:
Ahh so you recursively call borrow_mut down, in the tuple impl... you can do that with type system only.

Reply:
Let see that then?

Reply:
I don't wannt toooooo I just wanna merge thissss 😭